File: leetcode/golang/230_kth_smallest_element_in_bst.go
type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}

func kthSmallest(root *TreeNode, k int) int {
    count, res := 0, 0;
    var dfs func(root *TreeNode)
    dfs = func(root *TreeNode) {
        if root == nil || count >= k {
            return;
        }
        dfs(root.Left)
        count++
        if count == k {
            res = root.Val
            return
        }
        dfs(root.Right)
    }
    dfs(root)
    return res
}

File: leetcode/c++/230_kth_smallest_element_in_bst.cpp
// https://leetcode.cn/problems/kth-smallest-element-in-a-bst/description

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        this->k = k;
        dfs(root);
        return res;
    }
private:
    int res, k;
    void dfs(TreeNode* root) {
        if (root == nullptr) return;
        dfs(root->left);
        if (k==0) return;
        if (--k == 0) res = root->val;
        dfs(root->right);
    }
};

// 时间复杂度和空间复杂度最差为O(n)
